Enhance Executive Audit Validations (#338)
makerdao/governance-manual: governance/executive-audit.md

In the above example, the comparison is between this [Github file](https://github.com/makerdao/spells-mainnet/blob/master/archive/2021-09-24-DssSpell.sol) against the corresponding [Etherscan contract code](https://etherscan.io/address/0x0ed5a04DdE29f90bB00529608D3f17C1ffF778A0#code). 

### Validate the tests pass

> Validate that all of the forge tests pass on the spell's respective branch (see above discussion for determining the spell's branch). You can execute the tests using the command: `make test`.

Reviewer: Just to clarify, in this section, are you suggesting that we perform the tests locally using dappTools, just as the developers do? Or should we simply review what the developers have commented?

Author: Correct. There is existing documentation that requires this, althought it does require local setup of the dev environment, in an opsec friendly manner.

Reviewer: Now I understand. So, we could be a bit more specific. I'll leave you a suggestion, but feel free to modify it and elaborate further.
```suggestion
Validate that all of the forge tests pass on the spell's respective branch (see above discussion for determining the spell's branch). You can execute the tests by setting up a local environment, using dappTools, and running the command: `make test`.
```

```
"2021-09-24 MakerDAO Executive Spell | Hash: 0x655e71cb00b63a11c14db615d43d3e59202ff9d2b2bc6a6a03de42e258bd1be8";
````

To validate that the governance portal copy on vote.makerdao.com which is reflected on [Github](https://github.com/makerdao/community/tree/master/governance/votes) has the same hash as is stored in the executive spell, the user can copy and paste the body of the github text into this online [Github tool](https://emn178.github.io/online-tools/keccak_256.html) to generate and match the hash.
Make sure to replace the contract address field with the hard coded string
`$spell_address`, for example:
```diff
5c5
< address: "0x70254BD530684CF4a6323F51098FA39AAE6130b6"
---
> address: "$spell_address"
```

Alternatively, the user can generate this hash run the following seth command:

```
seth keccak -- "$(wget https://raw.githubusercontent.com/makerdao/community/f33870a7938c1842e8467226f8007a2d47f9ddeb/governance/votes/Executive%20vote%20-%20October%208%2C%202021.md -q -O - 2>/dev/null)"
```

> Where the referenced commit has the executive document with the `$spell_address`

Reviewer: ```suggestion
Where the URL references the commit that contains the executive document with the $spell_address template variable (which has not yet been replaced with the contract address).
```

> template variable (not replaced with the contract address).

Reviewer: There is a line break that divides the text, so it would be better to remove it if you approve my suggestion above. Otherwise, we should only remove the line break.
```suggestion
```

If the hash does not match, it would be prudent to question the spell contents and the copy that was referenced or changed.

## Anatomy of an Executive
An executive spell is broken into a number of different parts, including; spell libraries, spell actions, and constructors, which will be reviewed below.

### Spell Library
MakerDAO's executive spells have evolved over time towards using libraries in an effort to automate the process of making repetitive updates and changes to protocol parameters. Libraries reduce the chance of errors by reusing or building upon prior tested and verified code.

In auditing the executive it is first necessary to check that the spell uses the dss-exec-lib library and secondly, that the library is the official and legitimate library. This can be confirmed by looking at the spell copy in Etherscan.

> Presently, this library version is `v0.0.9` deployed at address `0x8de6ddbcd5053d32292aaa0d2105a32d108484a6`.

Reviewer: ```suggestion
Up until the last revision of this document, this library version is `v0.0.9` deployed at address `0x8de6ddbcd5053d32292aaa0d2105a32d108484a6`. It is possible to verify this from the [dss-exec-lib repository](https://github.com/makerdao/dss-exec-lib).```

Validating all rate adjustments can be done the same way. For more information on the rates module, please visit the [developer guide](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md). For easy reference, common pre-computed rates can also be viewed at the following [ipfs link](
https://ipfs.io/ipfs/QmefQMseb3AiTapiAKKexdKHig8wroKuZbmLtPLv4u2YwW).

> Rate unit constants (e.g., `MILLION`, `WAD`, `RAD`, `RAY`) should be validated. These constants are always defined the same, and can be validated by referencing the spell archive. The correct usage of these units and denominations can be determined by looking at the function argument names, or by referencing the same previous usage from the spell archive.

Reviewer: To validate the constants WAD, RAD, and RAY, the [developer guide](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md#fixed-point-numbers) can be used.
```suggestion
Rate unit constants (e.g., `MILLION`, `WAD`, `RAD`, `RAY`) should be validated. These constants are always defined the same, and the Fixed Point Numbers (`WAD`, `RAD`, `RAY`) can be validated using the previously mentioned [developer guide](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md#fixed-point-numbers). The correct usage of these units and denominations can be determined by looking at the function argument names, or by referencing the same previous usage from the spell archive.
```

You can validate that the spell was not deployed using `CREATE2` by looking at ethscan and clicking on the `createtxn` transaction. If the `To` field displays as "[0x... Created]" then it was not created using `CREATE2`. For example, see [this](https://etherscan.io/tx/0xc90fb0bd8979e8cbedf5879d736fe2f933103e18b152eb96f1db47a91b23fa23) transaction.
For example of a `CREATE2` transaction see [this](https://etherscan.io/tx/0x1888125e44400ae7d8bb074f65e88d6a3f060f7ccc6c38a6f4bc730c98f0ed33). Note the `To` field format, and click "Internal Txns" to see `create2_0_1_1`.

### Validate compiler optimizations
The solidity compiler optimizations should be `No with 200 runs`. You can view this on Etherscan under "Contract > Optimizations Enabled". This is to avoid solidity compiler exploits which can be triggered using only a specific number of runs. The only exception to this rule is for very large spells the contract size may be prohibitive and require optimizations to be enabled. In this case, the choice of the number of runs should be an expected value such as `1` or `100`.

### Validate the EVM version and license file

> The solidity compiler settings should be `default evmVersion`. You can view this on Etherscan under "Contract > Other Settings." If this is not the case, there should be an explanation as to why.

Reviewer: I have a doubt regarding this part, I understand the point, but I have seen some spells with the compiler setting London EvmVersion. I'm not sure if the developers already allow that variation. We should check on it.

Author: Thank you for checking. There is existing executive audit documentation that requires `default`, but about 50% of the spells I've validated so far have either `default` or `london`. It would be good to reach a resolution on the policy, so we can either update the documentation or require it in the development process.

Reviewer: I agree, I will discuss it with the team to gather other opinions and will let you know.

The license file should be `GNU AGPLv3`, which is displayed under the same section on Etherscan.

### Validate the solc compiler version matches `spells-mainnet`
The compiler version for the deployed spell should match the version specified on that spell's branch in the `spells-mainnet` repo. This is defined at the top of the `Makefile` under `--use solc:0.8.16 build`. The commit hash in the version does not need to be validated.

> ### Spell actions validations and provenance

Reviewer: This could perhaps be added to **Spell Actions** since it is part of the exhaustive control. What do you think?

Author: As in, move it out of the `Non-Exhaustive Checklist` sectoin, and have it be right before that section (and after `Office Hours`?)

Reviewer: It could be in the place you mentioned, or even within the same **Spell Actions** section, at the end of it, since we are still following the actions.
